scripts/update-spec-links.sh

```diff
@@ -0,0 +1,718 @@
+        ref="$1"
+      elif [[ -d "$1" ]]; then
+        dirs_in+=( "$1" )
+      elif [[ -f "$1" ]]; then
+        files_in+=( "$1" )
+      else
+        printf 'File or directory not found: %s\n' "$1"
+        exit 1
+      fi
+      ;;
+  esac
+  shift
+done
+
+if [[ -z "$ref" ]]; then
+  printf '%s\n' "$simple_usage"
+  printf 'No ref provided.\n'
+  exit 1
+fi
+
+if [[ -n "$verbose" ]]; then
+  if [[ "${#files_in[@]}" -eq '0' ]]; then
+    printf 'No files provided.\n'
+  else
+    printf 'Files provided (%d):\n' "${#files_in[@]}"
+    printf '  %s\n' "${files_in[@]}"
+  fi
+  if [[ "${#dirs_in[@]}" -eq '0' ]]; then
+    printf 'No directories provided.\n'
+  else
+    printf 'Directories provided (%d):\n' "${#dirs_in[@]}"
+    printf '  %s\n' "${dirs_in[@]}"
+  fi
+fi
+
+if [[ "$source_ref" == '--ref' ]]; then
+  source_ref="$ref"
+fi
+
+if [[ ! -f "$identify_messages_awk" ]]; then
+  printf 'Could not find the identify-messages awk script: %s\n' "$identify_messages_awk"
+  stop_early='1'
+fi
+if [[ ! -f "$identify_endpoints_awk" ]]; then
+  printf 'Could not find the identify-endpoints awk script: %s\n' "$identify_endpoints_awk"
+  stop_early='1'
+fi
+if [[ ! -f "$identify_links_awk" ]]; then
+  printf 'Could not find the identify-links awk script: %s\n' "$identify_links_awk"
+  stop_early='1'
+fi
+
+if [[ -n "$stop_early" ]]; then
+  exit 1
+fi
+
+
+########################################################################################################################
+########################################  Identify files that will be updated.  ########################################
+########################################################################################################################
+
+link_prefix="+++ https://github.com/provenance-io/provenance/blob/$ref/"
+link_rx='^\+\+\+ https://github\.com/provenance-io/provenance.*/proto/.*\.proto'
+link_rx_esc="$( sed -E 's|\\|\\\\|g;' <<< "$link_rx" )"
+declare files=()
+
+if [[ "${#files_in[@]}" -eq '0' && "${#dirs_in[@]}" -eq '0' ]]; then
+  dirs_in+=( '.' )
+fi
+
+i=0
+for file in "${files_in[@]}"; do
+  i=$(( i + 1 ))
+  [[ -n "$verbose" ]] && printf '[%d/%d]: Checking file: %s ... ' "$i" "${#files_in[@]}" "$file"
+  if grep -E --no-messages --binary-file=without-match --silent "$link_rx" "$file" > /dev/null 2>&1; then
+    files+=( "$file" )
+    [[ -n "$verbose" ]] && printf 'File has links.\n'
+  else
+    if [[ -n "$verbose" ]]; then
+      printf 'No links found.\n'
+    elif [[ -z "$quiet" ]]; then
+     printf 'File does not have any links: %s\n' "$file"
+   fi
+  fi
+done
+
+i=0
+for folder in "${dirs_in[@]}"; do
+  i=$(( i + 1 ))
+  [[ -n "$verbose" ]] && printf '[%d/%d]: Checking directory for files with links: %s ... ' "$i" "${#dirs_in[@]}" "$folder"
+  declare new_files=()
+  set -o noglob
+  IFS=$'\n' new_files+=( $( grep -E --files-with-matches --recursive --no-messages --binary-file=without-match "$link_rx" "$folder" | sort ) )
```

Comment:
Here again, `mapfile` is not something we can reliably expect environments to have.
Similarly, not all systems have a `read` utility that recognizes the `-a` flag.

Reply:
Continuing this conversation here: https://github.com/provenance-io/provenance/pull/2068#discussion_r1669060594 since you're just saying the same things.